### PR TITLE
Tray

### PR DESCRIPTION
**Pull request by willrogers**
@nickbattam @mfurseman first version.  Don't merge yet.

### CODE REVIEW

File: plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayApplicationWorkbenchWindowAdvisor.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                store.save();
+            } catch (IOException e) {
+                Plugin.getLogger().warning(Messages.TrayPreferences_saveFailed + e.getMessage());
+            }
+        }
+        return response;
+    }
+
+    @Override
+    public boolean preWindowShellClose() {
```
Comment (nickbattam) [inline plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayApplicationWorkbenchWindowAdvisor.java]: Each of the if blocks in this method end with a return. This means all the else blocks are redundant.

File: plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayIcon.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import org.eclipse.swt.widgets.Shell;
+import org.eclipse.swt.widgets.TrayItem;
+import org.eclipse.ui.IWorkbenchWindow;
+import org.eclipse.ui.PlatformUI;
+
+public class TrayIcon {
+
+    public static final String IMAGE = "icons/css.ico";
+    private TrayItem trayItem;
+    final private Image image = Activator.getImageDescriptor(IMAGE).createImage();
```
Comment (nickbattam) [inline plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayIcon.java]: For consistency this should be `private final Image`

File: plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayApplicationWorkbenchWindowAdvisor.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        if (PlatformUI.getWorkbench().getWorkbenchWindowCount() > 1) {
+            return super.preWindowShellClose();
+        } else {
+            IPreferencesService prefs = Platform.getPreferencesService();
+            String minPref = prefs.getString(Plugin.ID, TrayIconPreferencePage.MINIMIZE_TO_TRAY, null, null);
+            if (trayIcon.isMinimized() || minPref.equals(MessageDialogWithToggle.NEVER)) {
+                return super.preWindowShellClose();
+            } else {
+                if (minPref.equals(MessageDialogWithToggle.PROMPT)) {
+                    int response = prompt();
```
Comment (nickbattam) [inline plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayApplicationWorkbenchWindowAdvisor.java]: would this be clearer as a `switch` statement? You have an integer variable to switch on

File: plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayApplicationWorkbenchWindowAdvisor.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+    public int prompt() {
+        Shell parent = PlatformUI.getWorkbench().getActiveWorkbenchWindow().getShell();
+        ScopedPreferenceStore store = new ScopedPreferenceStore(InstanceScope.INSTANCE, Plugin.ID);
+        MessageDialogWithToggle dialog = new MessageDialogWithToggle(parent, Messages.TrayDialog_title, null,
+                Messages.TrayDialog_question, MessageDialog.QUESTION,
+                BUTTON_LABELS, 2, Messages.TrayDialog_rememberDecision, false);
+        dialog.open();
+        int response = dialog.getReturnCode();
+        if (dialog.getToggleState()) {
```
Comment (nickbattam) [inline plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayApplicationWorkbenchWindowAdvisor.java]: what does the `toggleState` represent here?

File: plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayIcon.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    private Menu menu;
+    private MenuItem open;
+    private MenuItem exit;
+    private IWorkbenchWindow window;
+
+    private boolean minimized;
+
+    private void dispose() {
+        trayItem.dispose();
```
Comment (nickbattam) [inline plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayIcon.java]: do you need to dispose the `image` too?

File: plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayIcon.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        shell.setFocus();
+        shell.setMinimized(false);
+        shell.layout();
+    }
+
+    public void minimize() {
+        trayItem = new TrayItem(Display.getCurrent().getSystemTray(), SWT.NONE);
+        trayItem.setImage(image);
+        trayItem.setToolTipText(Messages.TrayIcon_tooltip);
+        // There should be exactly one workbench window when this is being called.
```
Comment (nickbattam) [inline plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayIcon.java]: what happens if there isn't exactly one window?

File: plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayWorkbench.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+import org.csstudio.startup.application.OpenDocumentEventProcessor;
+import org.csstudio.startup.module.WorkbenchExtPoint;
+import org.csstudio.utility.product.Workbench;
+import org.eclipse.ui.application.WorkbenchAdvisor;
+
+public class TrayWorkbench extends Workbench implements WorkbenchExtPoint {
+
+    /**
+     * Creates a workbench advisor to be used by the created workbench. This methods allows
```
Comment (nickbattam) [inline plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayWorkbench.java]: is this a cut+paste comment from somewhere else?

File: plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayIcon.java (hunk 5 above)
Comment (willrogers) [inline plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayIcon.java]: We're using the same image for the lifetime of the application.  We may want to do the same for the whole tray icon and just hide it or show it as appropriate.

Comment (nickbattam) [inline plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayIcon.java]: I thought that was the case, but thought it wouldn't hurt to check

File: plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayApplicationWorkbenchWindowAdvisor.java (hunk 3 above)
Comment (willrogers) [inline plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayApplicationWorkbenchWindowAdvisor.java]: I had a little go and I couldn't make it look clearer.

File: plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayApplicationWorkbenchWindowAdvisor.java (hunk 1 above)
Comment (willrogers) [inline plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayApplicationWorkbenchWindowAdvisor.java]: I'm not sure what you mean.

**Comment by willrogers** [comment]
@nickbattam could you pick this up for the time being?  I'm looking very busy today and tomorrow.

There are two questions: the simpler one is the method you commented on.

The main one is what to do if the preference says 'start minimized' but more than one window was open when CS-Studio was closed so it would usually start with multiple windows.

File: plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayIcon.java [hunk 8]
```diff
@@ -114,7 +106,7 @@ public void handleEvent(Event event) {
 
     private Menu createPopupMenu() {
         // Create a Menu
-        Menu popupMenu = new Menu(window.getShell(), SWT.POP_UP);
+        Menu popupMenu = new Menu(windows[0].getShell(), SWT.POP_UP);
```
Comment (nickbattam) [inline plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayIcon.java]: Does this need better null checking?
`windows` will be set here; are there any use-cases in which `getWorkbenchWindows` could return an empty list?

**Comment by nickbattam** [comment]
@willrogers

File: plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayApplicationWorkbenchWindowAdvisor.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    /**
+     * If this is the last shell prompt handle minimize to tray option. If the
+     * 'never' preference isn't set, prompt the user for an action.
+     *
+     * @see org.eclipse.ui.application.WorkbenchWindowAdvisor#preWindowShellClose
+     */
+    @Override
+    public boolean preWindowShellClose() {
```
Comment (willrogers) [inline plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayApplicationWorkbenchWindowAdvisor.java]: This method still needs comments to explain what's happening, I think.

**Comment by willrogers** [comment]
Yes, minimising all windows works for me.

Do we have anything left before we merge into integration?

File: plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayApplicationWorkbenchWindowAdvisor.java (hunk 9 above)
Comment (nickbattam) [inline plugins/org.csstudio.trayicon/src/org/csstudio/trayicon/TrayApplicationWorkbenchWindowAdvisor.java]: I'll do another rewrite to see if that can improves clarity

**Comment by nickbattam** [comment]
@willrogers Rewritten preWindowShellClose a few times; I think the final commit is the easiest to understand.

**Comment by willrogers** [comment]
Yes, much clearer now.  Shall I merge for the nightly build?

**Pull request merged**